k2/csrc/rnnt_decode.cu

```diff
@@ -0,0 +1,732 @@
+/**
+ * Copyright      2022  Xiaomi Corporation (authors: Daniel Povey, Wei kang)
+ *
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include <limits>
+#include <memory>
+#include <vector>
+
+#include "k2/csrc/fsa.h"
+#include "k2/csrc/macros.h"
+#include "k2/csrc/ragged_ops.h"
+#include "k2/csrc/rnnt_decode.h"
+
+namespace k2 {
+namespace rnnt_decoding {
+
+std::shared_ptr<RnntDecodingStream> CreateStream(
+    const std::shared_ptr<Fsa> &graph) {
+  K2_CHECK_EQ(graph->shape.NumAxes(), 2);
+  ContextPtr &c = graph->shape.Context();
+  RnntDecodingStream stream;
+  stream.graph = graph;
+  stream.num_graph_states = graph->shape.Dim0();
+  // initialize to start state
+  stream.states = Ragged<int64_t>(RegularRaggedShape(c, 1, 1),
+                                  Array1<int64_t>(c, std::vector<int64_t>{0}));
+  stream.scores = Ragged<double>(RegularRaggedShape(c, 1, 1),
+                                 Array1<double>(c, std::vector<double>{0.0}));
+  return std::make_shared<RnntDecodingStream>(stream);
+}
+
+RnntDecodingStreams::RnntDecodingStreams(
+    std::vector<std::shared_ptr<RnntDecodingStream>> &srcs,
+    const RnntDecodingConfig &config)
+    : attached_(true), srcs_(srcs), num_streams_(srcs.size()), config_(config) {
```

Comment:
Initialization order should match the definition order to avoid complaints from the compiler.